File: Swift/ExampleTests/common/DiscountInputString.swift
public struct DiscountInputString: CustomStringConvertible, Equatable {
    public let totalPrice: String
    public let discount: String
    public let notes: String

    public init(totalPrice: String, discount: String, notes: String) {
        self.totalPrice = totalPrice
        self.discount = discount
        self.notes = notes
    }

    public init(fromArray v: [String]) {
        self.totalPrice = v.count > 0 ? v[0] : ""
        self.discount = v.count > 1 ? v[1] : ""
        self.notes = v.count > 2 ? v[2] : ""
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    public static func fromText(_ text: String) -> DiscountInputString {
        let parts = Tokens.require(text, 3, "DiscountInput")
        return DiscountInputString(totalPrice: parts[0], discount: parts[1], notes: parts[2])
    }

    public var description: String {
        return Tokens.token(totalPrice) + " " + Tokens.token(discount) + " " + Tokens.token(notes)
    }

    public static let dncString = "?DNC?"

    public static func == (a: DiscountInputString, b: DiscountInputString) -> Bool {
        return (a.totalPrice == dncString || b.totalPrice == dncString || a.totalPrice == b.totalPrice)
            && (a.discount == dncString || b.discount == dncString || a.discount == b.discount)
            && (a.notes == dncString || b.notes == dncString || a.notes == b.notes)
    }
}

File: Python/ExampleTests/common/discount_input_string.py
from . import tokens
DNC_STRING = '?DNC?'


class DiscountInputString:
    def __init__(self, total_price: str = '', discount: str = '', notes: str = ''):
        self.total_price = total_price
        self.discount = discount
        self.notes = notes

    @classmethod
    def from_list(cls, values):
        v = list(values)
        return cls(
            v[0] if len(v) > 0 else '',
            v[1] if len(v) > 1 else '',
            v[2] if len(v) > 2 else ''
        )

    @classmethod
    def from_text(cls, text):
        """Builds from the text form, e.g. Money as '25 USD'."""
        parts = tokens.require(text, 3, 'DiscountInput')
        return cls(
            parts[0],
            parts[1],
            parts[2]
        )

    def __str__(self):
        return (tokens.token(self.total_price) + ' ' +
                tokens.token(self.discount) + ' ' +
                tokens.token(self.notes))

    def _key(self):
        return (self.total_price, self.discount, self.notes)

    def __eq__(self, other):
        if not isinstance(other, DiscountInputString):
            return NotImplemented
        return all(a == b or a == DNC_STRING or b == DNC_STRING
                   for a, b in zip(self._key(), other._key()))

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'DiscountInputString({self})'
